### PR TITLE
DOC: Added an example of pitfalls when using astype

### LINKED ISSUE #13260: DOC: Strange behavior when combining astype and loc
I am trying to use `astype` together with `loc` to convert just a subset of columns to a specified dtype. However, I can't understand what is actually happening. Is this the expected behavior when combining 
`astype` and `loc`?

#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
import numpy as np

df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})

print df.loc[:, ['a', 'b']].astype(np.uint8).dtypes

df.loc[:, ['a', 'b']] = df.loc[:, ['a', 'b']].astype(np.uint8)
print df.dtypes
```

Output:
```
>>> df.loc[:, ['a', 'b']].astype(np.uint8).dtypes
a    uint8
b    uint8
dtype: object
>>> df.loc[:, ['a', 'b']] = df.loc[:, ['a', 'b']].astype(np.uint8)
>>> df.dtypes
a    int64
b    int64
c    int64
dtype: object
```

#### Expected Output
```
>>> df.loc[:, ['a', 'b']].astype(np.uint8).dtypes
a    uint8
b    uint8
dtype: object
>>> df.loc[:, ['a', 'b']] = df.loc[:, ['a', 'b']].astype(np.uint8)
>>> df.dtypes
a    uint8
b    uint8
c    int64
dtype: object
```

#### output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.0-36-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: None
pip: 8.1.1
setuptools: 20.6.7
Cython: None
numpy: 1.11.0
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: None
patsy: None
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None

### PR DESCRIPTION
**Pull request by pfrcks**
- [x] closes #13260 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/source/basics.rst [hunk 1]
```diff
@@ -1726,6 +1726,23 @@ then the more *general* one will be used as the result of the operation.
    # conversion of dtypes
    df3.astype('float32').dtypes
 
+When trying to convert a subset of columns to a specified type using :meth:`~DataFrame.astype`  and :meth:`~numpy.ndarray.loc`, utilizing **:** as mask, upcasting occurs.
```
Comment (jorisvandenbossche) [inline doc/source/basics.rst]: ```:meth:`~numpy.ndarray.loc` ``` -> ``` :meth:`~DataFrame.loc`   ```

File: doc/source/basics.rst [hunk 2]
```diff
@@ -1726,6 +1726,23 @@ then the more *general* one will be used as the result of the operation.
...
+:meth:`~numpy.ndarray.loc` tries to fit in what we are assigning to the current dtypes, while [ ] will overwrite them taking the dtype from the right hand side.
+
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
+   print df.loc[:, ['a', 'b']].astype(np.uint8).dtypes
```
Comment (jorisvandenbossche) [inline doc/source/basics.rst]: 'print' is not needed

**Comment by jorisvandenbossche** [comment]
@pfrcks Thanks for the contribution!

Personally, I would reverse the logic in the explanation, by first saying how you should do it, and only then what does not work.

Eg in the sense of "When you want to astype subset of columns -> you have to do it like this: ... -> attention, you cannot use loc in this case like ... as this upcasts"

**Comment by pfrcks** [comment]
@jorisvandenbossche I have made the cleanup changes you recommended. Do I need to change the order of the wrong and right way of doing things?
Wrong followed by right way of doing things also sends a clear message I guess. We see how the thing is wrong, then we see how to correct or implement it properly.

File: doc/source/basics.rst [hunk 3]
```diff
@@ -1726,6 +1726,24 @@ then the more *general* one will be used as the result of the operation.
...
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
+   df.loc[:, ['a', 'b']].astype(np.uint8).dtypes
+   df.loc[:, ['a', 'b']] = df.loc[:, ['a', 'b']].astype(np.uint8)
+
+To avoid this please take the following approach.
+
+.. ipython:: python
+
```
Comment (jreback) [inline doc/source/basics.rst]: flip these around showing the correct method of assignment using astype.

then you can put a ``::note`` for the explanation. People reading this just want the answer; *some* want the explanation, so highliting that differently helps.

**Comment by pfrcks** [comment]
Restructured the documentation.

File: doc/source/basics.rst [hunk 4]
```diff
@@ -1726,6 +1726,28 @@ then the more *general* one will be used as the result of the operation.
...
+   df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
+   df[['a','b']] = df[['a','b']].astype(np.uint8)
+   df
+   df.dtypes
+
+.. note::
+
+    When trying to convert a subset of columns to a specified type using :meth:`~DataFrame.astype`  and :meth:`~DataFrame.loc`, utilizing **:** as mask, upcasting occurs.
+
+    :meth:`~DataFrame.loc` tries to fit in what we are assigning to the current dtypes, while [ ] will overwrite them taking the dtype from the right hand side. Therefore the following piece of code produces the unintended result.
```
Comment (jreback) [inline doc/source/basics.rst]: double backticks around ``[]``

File: doc/source/basics.rst [hunk 5]
```diff
@@ -1726,6 +1726,28 @@ then the more *general* one will be used as the result of the operation.
...
+
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
```
Comment (jreback) [inline doc/source/basics.rst]: you already created the df, no need to recreate

File: doc/source/basics.rst [hunk 6]
```diff
@@ -1726,6 +1726,28 @@ then the more *general* one will be used as the result of the operation.
    # conversion of dtypes
    df3.astype('float32').dtypes
 
+Convert a subset of columns to a specified type using :meth:`~DataFrame.astype`
+
+.. ipython:: python
+
+   df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
```
Comment (jreback) [inline doc/source/basics.rst]: call this ``dft`` or something besides df (as that's used a lot)

File: doc/source/basics.rst (hunk 5 above)
Comment (pfrcks) [inline doc/source/basics.rst]: df get's changed in the code example before.

File: doc/source/basics.rst [hunk 7]
```diff
@@ -1726,6 +1726,28 @@ then the more *general* one will be used as the result of the operation.
...
+.. ipython:: python
+
+   dft = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6], 'c': [7, 8, 9]})
+   dft[['a','b']] = dft[['a','b']].astype(np.uint8)
+   dft
+   dft.dtypes
+
+.. note::
+
+    When trying to convert a subset of columns to a specified type using :meth:`~DataFrame.astype`  and :meth:`~DataFrame.loc`, utilizing **:** as mask, upcasting occurs.
```
Comment (jorisvandenbossche) [inline doc/source/basics.rst]: `utilizing **:** as mask` is not needed I think (in the text), as you cannot assign columns without it.

**Comment by pfrcks** [comment]
@jorisvandenbossche Made the suggested changes. Hope it is fine now.

File: doc/source/basics.rst [hunk 8]
```diff
@@ -1726,6 +1726,28 @@ then the more *general* one will be used as the result of the operation.
...
+   dft
+   dft.dtypes
+
+.. note::
+
+    When trying to convert a subset of columns to a specified type using :meth:`~DataFrame.astype`  and :meth:`~DataFrame.loc`, upcasting occurs.
+
+    :meth:`~DataFrame.loc` tries to fit in what we are assigning to the current dtypes, while ``[]`` will overwrite them taking the dtype from the right hand side. Therefore the following piece of code produces the unintended result.
+
+.. ipython:: python
```
Comment (jorisvandenbossche) [inline doc/source/basics.rst]: Can you indent this one (and the following lines) to the same level as 'When trying to convert ...', so the code example is also included in the 'note' frame

**Comment by jorisvandenbossche** [comment]
Only a small last comment, for the rest looks good!

**Comment by pfrcks** [comment]
@jorisvandenbossche I hope that does the trick ;)

**Comment by jorisvandenbossche** [comment]
@pfrcks Thanks!

**Pull request closed without merging**